Add definition of recursive gossip (#254)
ethereum/portal-network-specs: state-network.md

Any node in the trie that represents a value stored in the trie.  The nodes in the diagram at level 4 are leaf nodes.

#### *"merkle proof"* or *"proof"*

> A collection of nodes from the trie sufficient to recompute the state root and proove that one or more nodes are part of the trie defined by that state root.

Reviewer: ```suggestion
A collection of nodes from the trie sufficient to recompute the state root and prove that one or more nodes are part of the trie defined by that state root.
```

> > A proof is considered to be "minimal" if it contains only the minimumm set of trie nodes needed to recompute the state root.

Reviewer: ```suggestion
> A proof is considered to be "minimal" if it contains only the minimum set of trie nodes needed to recompute the state root.
```

### Overview

The goal of the "recursive gossip" mechanism is to reduce the burden of
responsibility placed on bridge nodes for injecting new state data into the
network while simultaniously spreading the responsibility for gossiping new
state data across the nodes in the network.

At each block we construct a proof against the new state root which contains
all of the state changes which occured within that block.  

> This proof would *explicitely* contains a mixed set of leaf and intermediate
> nodes, as well as implicitely contains a set of intermediate nodes which can be
> computed from the nodes that are part of the proof.

Reviewer: ```suggestion
This proof contains *explicitly* a mixed set of leaf and intermediate
nodes, as well as implicitly a set of intermediate nodes which can be
computed from the nodes that are part of the proof.
```

```
EXAMPLE: Recursive Gossip Inception

0:                           A*
                            / \
                          /     \
                        /         \
                      /             \
                    /                 \
                  /                     \
                /                         \
1:             B*                          C
             /   \
           /       \
         /           \
2:      D             E*
                     / \
                    /   \
3:                 F     G*
                        / \
4:                     H*  I

- "*" denotes trie node modified
```

In the example proof diagramed here we can see a *leaf* node `H` which
represents the only modified leaf state in this proof.  Note that all of the
nodes along the path leading to `H` are also modified.

The *minimal* proof would contain the nodes `[D, F, H, I, C]`

> The bridge node would search the DHT for nodes that are *inderested* in storing

Reviewer: ```suggestion
The bridge node would search the DHT for nodes that are *interested* in storing
```

the node `H` and gossip this proof to those nodes.

The recipients of this gossip are then responsible for gossiping the parent
intermediate node `G`.  To do so, they would strip off the `H` and `I` nodes
from this proof resulting in the following:

```
EXAMPLE: Recursive Gossip Round 1

0:                           A*
                            / \
                          /     \
                        /         \
                      /             \
                    /                 \
                  /                     \
                /                         \
1:             B*                          C
             /   \
           /       \
         /           \
2:      D             E*
                     / \
                    /   \
3:                 F     G*

4:
```

At this stage, the minimal proof for `G` would be `[D, F, G, C]`.  The nodes

> which received the initial gossip message for `H` would constructe this proof
> by removing the un-necessary nodes, after which they would searcch the DHT for

Reviewer: ```suggestion
which received the initial gossip message for `H` would construct this proof
by removing the un-necessary nodes, after which they would search the DHT for
```

nodes that are interested in `F` and gossip this proof to them..

> The receipients of that gossip are then responsible for gossiping the parent

Reviewer: ```suggestion
The recipients of that gossip are then responsible for gossiping the parent
```
